File: plugins/js-ts/src/bulkIndex.ts
// Initial bulk index: walks a project's JS/TS/JSX/TSX files, runs the
// existing tree-sitter extraction (extract.ts) on each one, and streams the
// resulting nodes/edges out as NDJSON (one compact JSON object per line)
// matching core's wire contract (core/src/protocol/types.rs - `WireNode` /
// `WireEdge`). Driven by index.ts's one-shot `--bulk-index` mode, whose whole
// stdout is this stream - never by the long-lived control plane, which speaks
// framed request/response instead.

import * as fs from "node:fs/promises";
import type { Dirent } from "node:fs";
import * as path from "node:path";

import { extractFile, isSupportedFile, type ExtractedEdge, type ExtractedNode } from "./extract";
import {
  HARD_EXCLUDED_DIRS,
  isIgnoredByLayers,
  loadGitignoreLayer,
  toPosixPath,
  type GitignoreLayer,
} from "./ignorePolicy";
import { createProjectResolver } from "./resolve";
import { canonicalizeProjectRoot, createSymlinkGuard, type SymlinkGuard } from "./symlinks";

/**
 * Where bulk-indexed NDJSON lines go. A callback mirrors the simplest test
 * seam; a `WritableStream` mirrors `writeMessage(stream, ...)` in
 * jsonrpc.ts so a future caller can pass `process.stdout` directly.
 */
export type NdjsonSink = NodeJS.WritableStream | ((line: string) => void);

/** Returns false when a stream sink is already over its high-water mark,
 * i.e. the consumer is behind and the caller should let it catch up. */
function emitLine(sink: NdjsonSink, value: unknown): boolean {
  const line = JSON.stringify(value);
  if (typeof sink === "function") {
    sink(line);
    return true;
  }
  return sink.write(line + "\n");
}

/**
 * Parks until a stream sink has flushed enough to accept more, so a project
 * bigger than the pipe buffer queues in the *consumer's* pace rather than in
 * this process's memory. Core reads this stream while committing each batch
 * to SQLite (`daemon::bulk_index`), so it is routinely slower than the walk;
 * without this, `bulkIndexProject`'s bounded-memory promise would hold for
 * extraction but not for output.
 */
function waitForDrain(sink: NdjsonSink): Promise<void> {
  if (typeof sink === "function") return Promise.resolve();
  return new Promise((resolve) => {
    const settle = (): void => {
      sink.removeListener("drain", settle);
      sink.removeListener("close", settle);
      sink.removeListener("error", settle);
      resolve();
    };
    sink.once("drain", settle);
    // A consumer that died never drains: resuming on close/error lets the
    // next write fail loudly instead of parking here forever.
    sink.once("close", settle);
    sink.once("error", settle);
  });
}

// --- wire shape (core/src/protocol/types.rs) -----------------------------

export interface WirePosition {
  line: number;
  col: number;
}

export interface WireRange {
  start: WirePosition;
  end: WirePosition;
}

/** Mirrors core's `WireNode`. The one real transformation off `ExtractedNode`:
 * flat startLine/startCol/endLine/endCol become a nested `range`. */
export interface WireNode {
  id: string;
  kind: ExtractedNode["kind"];
  name: string;
  qualifiedName: string;
  filePath: string;
  range: WireRange;
  signature: string | null;
  exported: boolean;
  docComment: string | null;
  language: string;
  nativeKind: string | null;
  hasSyntaxErrors: boolean;
}

/** Identical shape to `ExtractedEdge` - passed through unchanged. */
export type WireEdge = ExtractedEdge;

export function toWireNode(node: ExtractedNode): WireNode {
  return {
    id: node.id,
    kind: node.kind,
    name: node.name,
    qualifiedName: node.qualifiedName,
    filePath: node.filePath,
    range: {
      start: { line: node.startLine, col: node.startCol },
      end: { line: node.endLine, col: node.endCol },
    },
    signature: node.signature ?? null,
    exported: node.exported,
    docComment: node.docComment ?? null,
    language: node.language,
    nativeKind: node.nativeKind ?? null,
    hasSyntaxErrors: node.hasSyntaxErrors,
  };
}

// --- gitignore-aware walk -------------------------------------------------
// The exclusion policy itself lives in ignorePolicy.ts, shared with
// resolve.ts so the two cannot disagree about what gets indexed; the symlink
// policy lives in symlinks.ts, shared with workspace.ts's glob expansion for
// the same reason.

/**
 * Walks `projectRoot` depth-first, yielding project-relative POSIX paths of
 * every JS/TS/JSX/TSX file that is not gitignored. `.git`, `node_modules`,
 * `dist`, and `.claude` are never descended into, regardless of `.gitignore`
 * contents. Entries within a directory are visited in sorted order for
 * deterministic output.
 *
 * Symlinked directories and files are **followed**, and yielded under the path
 * they were reached by rather than the path they resolve to - a symlinked
 * package is indexed like any other. What that costs is a guard
 * (`createSymlinkGuard`): a link onto one of its own ancestors is refused
 * instead of descended into forever, a second path onto an already-walked real
 * location is dropped so nothing is indexed twice, a link resolving outside the
 * project root is refused, and a dangling link is skipped. The guard belongs to
 * this one walk and is threaded through the recursion.
 */
export async function* walkProjectFiles(projectRoot: string): AsyncGenerator<string> {
  // Canonicalized once, because every guard comparison is made against it and a
  // project root reached through a symlink would otherwise be spelled two ways.
  const projectRootReal = canonicalizeProjectRoot(projectRoot);
  yield* walkDir(projectRootReal, projectRootReal, [], createSymlinkGuard(projectRootReal));
}

async function* walkDir(
  projectRoot: string,
  dir: string,
  layers: readonly GitignoreLayer[],
  guard: SymlinkGuard,
): AsyncGenerator<string> {
  const ownLayer = await loadGitignoreLayer(dir);
  const nextLayers = ownLayer ? [...layers, ownLayer] : layers;

  let entries: Dirent[];
  try {
    entries = await fs.readdir(dir, { withFileTypes: true });
  } catch {
    return; // directory vanished / unreadable - nothing to yield
  }
  entries.sort((a, b) => a.name.localeCompare(b.name));

  for (const entry of entries) {
    // By name, before anything is resolved: a hard-excluded name is excluded
    // whatever it turns out to be - including a symlink *called* node_modules -
    // and settling that first also saves the guard's realpath call for it.
    if (HARD_EXCLUDED_DIRS.has(entry.name)) continue;

    const resolved = guard(dir, entry);
    if (resolved === null) continue; // cycle, duplicate real path, escapes projectRoot, or dangling
    const { absPath: abs, isDirectory, isFile } = resolved;

    if (isDirectory) {
      if (isIgnoredByLayers(nextLayers, abs)) continue;
      yield* walkDir(projectRoot, abs, nextLayers, guard);
      continue;
    }

    if (!isFile) continue; // neither file nor directory - a socket, device, fifo
    // Anything with an unsupported extension - including tsconfig.json - is
    // never opened here, which is what keeps a malicious tsconfig
    // `compilerOptions.plugins` entry inert (see security.test.ts).
    if (!isSupportedFile(abs)) continue;
    if (isIgnoredByLayers(nextLayers, abs)) continue;
    yield toPosixPath(path.relative(projectRoot, abs));
  }
}

// --- bulk index ------------------------------------------------------------

export interface BulkIndexSummary {
  filesProcessed: number;
  nodesEmitted: number;
  edgesEmitted: number;
}

/**
 * Walks `projectRoot`'s supported source files (gitignore-aware, see
 * `walkProjectFiles`), extracts each one with the existing tree-sitter pass,
 * and writes its nodes then its edges to `sink` as compact NDJSON - one line
 * per node/edge, emitted as soon as that file's extraction completes. Files
 * are processed one at a time; nothing from a later file is read before an
 * earlier file's output has been written, so memory use stays bounded by a
 * single file's extraction result, never the whole project's - and a stream
 * sink that falls behind pauses the walk at the next file boundary
 * (`waitForDrain`) rather than letting the backlog accumulate here.
 *
 * A file with syntax errors still contributes whatever nodes/edges
 * tree-sitter's error-tolerant parse produced (see `ExtractedNode.hasSyntaxErrors`) -
 * only files `isSupportedFile` rejects, or gitignored files, are skipped.
 */
export async function bulkIndexProject(
  projectRoot: string,
  sink: NdjsonSink,
): Promise<BulkIndexSummary> {
  let filesProcessed = 0;
  let nodesEmitted = 0;
  let edgesEmitted = 0;

  // One resolver (and therefore one existence memo) for the whole walk: this
  // process is one-shot, so nothing can create or delete a file behind its
  // back while it runs, and the same handful of shared modules is otherwise
  // re-stat-ed once per importing file.
  const resolveSpecifier = createProjectResolver(projectRoot);

  for await (const relPath of walkProjectFiles(projectRoot)) {
    const absPath = path.join(projectRoot, relPath);
    let sourceText: string;
    try {
      sourceText = await fs.readFile(absPath, "utf8");
    } catch {
      continue; // vanished between walk and read - not this ticket's concern
    }

    let result;
    try {
      result = extractFile(relPath, sourceText, { resolveSpecifier });
    } catch {
      continue; // e.g. UnsupportedFileError should not happen post-walk-filter,
      // but a corrupt/oversized file must not abort the whole bulk index
    }
    filesProcessed += 1;

    let sinkReady = true;
    for (const node of result.nodes) {
      sinkReady = emitLine(sink, toWireNode(node));
      nodesEmitted += 1;
    }
    for (const edge of result.edges) {
      sinkReady = emitLine(sink, edge);
      edgesEmitted += 1;
    }
    // Only at a file boundary: a file's nodes and edges belong together (an
    // edge is only ever emitted between nodes of the file it came from), and
    // keeping them in one uninterrupted run is what lets a consumer commit
    // whatever it has at any point without ever holding a dangling edge.
    if (!sinkReady) await waitForDrain(sink);
  }

  return { filesProcessed, nodesEmitted, edgesEmitted };
}

File: core/src/daemon/bulk_index.rs
//! Cold-start bulk index: the one full walk of a project that gives a
//! never-indexed codebase a populated graph before its daemon answers
//! anything. The file watcher cannot do this: `notify` reports only changes
//! made while it runs.
//!
//! Each discovered language's plugin is spawned in its one-shot
//! `--bulk-index` mode, one process per language, from the manifest's own
//! `command`/`args`. Its stdout is the EOF-terminated NDJSON stream
//! `protocol::ndjson::NdjsonReader` consumes. Why a second process rather
//! than the control-plane pipe, and why one failed language fails the whole
//! walk: [ADR 0002](../../../docs/adr/0002-bulk-walk.md).

use std::collections::BTreeSet;
use std::io::{BufRead, BufReader};
use std::path::Path;
use std::process::{Command, Stdio};

use anyhow::{bail, Context, Result};

use crate::daemon::indexing_status::IndexingStatus;
use crate::daemon::manifest::{DiscoveredPlugins, PluginManifest};
use crate::daemon::plugin;
use crate::embedding::EmbeddingPipeline;
use crate::protocol::ndjson::{BulkItem, NdjsonReader};
use crate::storage::index_store::{IndexStore, Unit, Writer};
use crate::storage::schema;
use crate::storage::write::Diff;
use crate::watcher::apply::{to_edge_record, to_node_record};
use crate::watcher::staleness;

/// Puts the plugin in one-shot bulk-index mode; must stay in sync with
/// `BULK_INDEX_FLAG` in plugins/typescript/src/index.ts.
pub(crate) const BULK_INDEX_FLAG: &str = "--bulk-index";

/// Set to `1` on every bulk spawn, telling the plugin its stdin is a lifeline:
/// a pipe core holds open and never writes, whose EOF means core is gone, so
/// the walk should stop. Opt-in by the spawner, so a plugin run by an older
/// core or by hand with `< /dev/null` does not read an immediate EOF as "exit
/// before walking". Must stay in sync with the plugins' own copies
/// (`plugins/sdk/src/run.rs`, `plugins/go/main.go`,
/// `plugins/typescript/src/index.ts`) - see
/// `docs/architecture/plugin-lifetime.md` §2.
pub(crate) const BULK_STDIN_LIFELINE_ENV: &str = "G_MESH_BULK_STDIN_LIFELINE";

/// Nodes plus edges accumulated before a batch is committed: bounds both the
/// memory held before a write and the number of transactions.
const BATCH_ITEMS: usize = 2_000;

/// `NodeKind::File`'s storage spelling (`watcher::apply::to_node_record`).
const FILE_NODE_KIND: &str = "File";

/// Test-only: holds a finished walk open for this many milliseconds before
/// [`run`] returns, so the daemon has not yet recorded the walk as complete.
/// Real installs never set it.
///
/// Applied *after* everything is committed: a test can wait for the row it is
/// about to ask for and only then ask, so a "still indexing" answer proves the
/// indexing flag gates the response rather than an empty table.
pub const WALK_DELAY_ENV: &str = "G_MESH_BULK_INDEX_DELAY_MS";

/// Test-only: path whose *deletion* releases the finished walk, so completion
/// is an event the test causes rather than a duration (unlike
/// [`WALK_DELAY_ENV`]). [`HOLD_LOCK_FILE_ENV`] is the sibling knob that holds
/// the batch-commit lock itself open.
pub const WALK_HOLD_FILE_ENV: &str = "G_MESH_BULK_INDEX_HOLD_FILE";

#[derive(Debug, Default, PartialEq, Eq)]
pub struct BulkIndexSummary {
    pub nodes: usize,
    pub edges: usize,
    /// Lines that parsed as neither a node nor an edge. Skipped rather than
    /// fatal, matching `NdjsonReader`'s contract: one unreadable line costs
    /// one symbol, not the project's whole index.
    pub skipped_lines: usize,
    /// `IMPORTS` edges the post-walk linking pass repointed from a module
    /// placeholder onto the real file it names (`graph::imports`).
    pub linked_imports: usize,
    /// `CALLS`/`REFERENCES`/`SUPERTYPE_OF` edges the post-walk linking pass
    /// repointed from a pending-symbol placeholder onto the symbol another
    /// file exports (`graph::symbol_links`).
    pub linked_symbols: usize,
}

/// Everything one walk carries from language to language and batch to batch:
/// the store it writes, what it reports to, and what it accumulates.
pub(crate) struct WalkContext<'a> {
    pub(crate) store: &'a IndexStore,
    /// `None` for the structural-only cold-start walk (`embedding::backfill`
    /// fills the vectors afterwards); `Some` for
    /// `daemon::workspace_reindex`'s per-language re-walk.
    pub(crate) embedding: Option<&'a EmbeddingPipeline>,
    /// The daemon's walk counters; `None` when nobody is waiting on them.
    pub(crate) progress: Option<&'a IndexingStatus>,
    pub(crate) summary: BulkIndexSummary,
    /// The file paths of every `File` node ingested, when the caller records
    /// staleness baselines for them; `None` otherwise.
    pub(crate) walked_files: Option<BTreeSet<String>>,
}

impl<'a> WalkContext<'a> {
    /// A structural walk into `store` that reports nowhere and records no
    /// walked files; set the other fields with struct-update syntax.
    pub(crate) fn new(store: &'a IndexStore) -> Self {
        Self {
            store,
            embedding: None,
            progress: None,
            summary: BulkIndexSummary::default(),
            walked_files: None,
        }
    }
}

/// Walks `project_root` through every plugin `discovered` names - one
/// one-shot `--bulk-index` process per language, run one after another - and
/// commits everything each of them emits, returning only once every child has
/// exited and the last batch is durable. `daemon::run` turns that return into
/// the moment its `IndexingStatus` flips and tools start answering for real,
/// so "returned" has to mean "complete".
///
/// Batching is safe to cut anywhere in one language's stream even though
/// edges are foreign keys onto nodes: a plugin emits a file's nodes before
/// that same file's edges, and never an edge between files (see the
/// dangling-edge guard in extract.ts). Every cross-file edge appears only
/// afterwards, when `graph::imports` and `graph::symbol_links` link the walk's
/// placeholders - once, project-wide, after every language's stream is over,
/// because only then does every node that will exist, exist.
pub fn run(
    project_root: &Path,
    conn: &IndexStore,
    embedding: Option<&EmbeddingPipeline>,
    discovered: &DiscoveredPlugins,
) -> Result<BulkIndexSummary> {
    run_with_progress(project_root, conn, embedding, discovered, None)
}

/// [`run`], also reporting through `progress`'s walk counters: languages done
/// out of total, the language being walked, and items ingested so far.
pub fn run_with_progress(
    project_root: &Path,
    conn: &IndexStore,
    embedding: Option<&EmbeddingPipeline>,
    discovered: &DiscoveredPlugins,
    progress: Option<&IndexingStatus>,
) -> Result<BulkIndexSummary> {
    // Sorted so languages are walked, and a failure is named, in a
    // deterministic order. Ingestion order does not change the result.
    let mut manifests: Vec<&PluginManifest> = discovered.manifests.values().collect();
    manifests.sort_by(|a, b| a.language.cmp(&b.language));

    if let Some(progress) = progress {
        progress.start_walk_progress(u32::try_from(manifests.len()).unwrap_or(u32::MAX));
    }
    // Taken before the first plugin is spawned, so no plugin can have read a
    // file before it - `staleness::record_walk_baselines` relies on that.
    let walk_started = std::time::SystemTime::now();
    let mut ctx =
        WalkContext { embedding, progress, walked_files: Some(BTreeSet::new()), ..WalkContext::new(conn) };
    for manifest in manifests {
        if let Some(progress) = progress {
            progress.mark_language_started(&manifest.language);
        }
        // One language's failure fails the whole walk, even for a language
        // the project has no files of: an index that silently skipped a
        // language would look complete to every tool (ADR 0002).
        walk_one_language(project_root, manifest, &mut ctx)?;
        if let Some(progress) = progress {
            progress.mark_language_done();
        }
    }
    let WalkContext { mut summary, walked_files, .. } = ctx;
    let walked_files = walked_files.unwrap_or_default();

    // Once, after every language's stream: an import links only to a file
    // that is already a node, and a cross-file usage only once every language
    // that might define it has run.
    let links = conn.link_all()?;
    summary.linked_imports = links.imports;
    summary.linked_symbols = links.symbols;

    // A staleness baseline for every walked file, so the first query of an
    // untouched file takes `staleness::ensure_fresh`'s fast path instead of a
    // synchronous reindex. Written after linking, so a row only exists for a
    // complete graph, and before the walk is reported done, so no query races
    // it. Best-effort: without baselines the walk is still complete; its files
    // reindex on first touch.
    match staleness::record_walk_baselines(
        conn,
        project_root,
        walked_files.iter().map(String::as_str),
        walk_started,
    ) {
        Ok(baselines) => {
            if baselines.skipped > 0 {
                eprintln!(
                    "g-mesh: {} of {} walked files got no staleness baseline (modified during or just \
                     before the walk) - each reindexes on its first query",
                    baselines.skipped,
                    walked_files.len()
                );
            }
        }
        Err(err) => eprintln!(
            "g-mesh: could not record the walk's staleness baselines - every file reindexes on its first \
             query: {err:#}"
        ),
    }

    hold_the_walk_open_for_tests();
    Ok(summary)
}

/// Spawns `manifest`'s plugin in its one-shot `--bulk-index` mode for
/// `project_root` and folds everything it emits into `ctx`, returning only
/// once the child has exited and every batch it produced is durable. Also
/// called by `daemon::workspace_reindex` to re-walk one language after
/// deleting its rows.
///
/// Runs as one [`Unit::BulkWalk`]: its batch commits and its bookkeeping row
/// are the unit's steps.
pub(crate) fn walk_one_language(
    project_root: &Path,
    manifest: &PluginManifest,
    ctx: &mut WalkContext<'_>,
) -> Result<()> {
    let store = ctx.store;
    store.unit(Unit::BulkWalk, |store| walk_one_language_in(project_root, manifest, store, ctx))
}

fn walk_one_language_in(
    project_root: &Path,
    manifest: &PluginManifest,
    store: &mut Writer<'_>,
    ctx: &mut WalkContext<'_>,
) -> Result<()> {
    // The check `daemon::plugin::PluginState::spawn` makes too: an unbuilt
    // cargo-workspace plugin binary gets a message naming the build command
    // instead of a bare "No such file or directory".
    if let Some(hint) = plugin::missing_plugin_binary_hint(&manifest.command, &manifest.args) {
        bail!("failed to spawn the {} plugin's bulk index: {hint}", manifest.language);
    }

    let mut command = Command::new(&manifest.command);
    command
        .args(&manifest.args)
        .arg(BULK_INDEX_FLAG)
        .arg(project_root)
        // The lifeline: a pipe this process never writes to. It stays inside
        // `child` - never taken, never dropped early - so the plugin sees EOF
        // exactly when this process's end closes, which the kernel does even
        // on SIGKILL. `Child::wait` below closes it before waiting, which is
        // harmless: by then stdout has reached EOF, so the plugin is already
        // exiting. The env var arms the plugin's watcher.
        .stdin(Stdio::piped())
        .env(BULK_STDIN_LIFELINE_ENV, "1")
        .stdout(Stdio::piped())
        // Plugin logs are diagnostic only: they go wherever the daemon's own
        // stderr goes.
        .stderr(Stdio::inherit());
    let mut child = crate::process::spawn_serialized(&mut command).with_context(|| {
        format!(
            "failed to spawn the {} plugin's bulk index ({})",
            manifest.language,
            manifest.command.display()
        )
    })?;

    let stdout = child.stdout.take().context("bulk-index plugin process has no stdout")?;

    if let Err(err) = ingest_in(BufReader::new(stdout), store, ctx) {
        // Nobody will read the rest of this walk: a plugin left writing into
        // an undrained pipe would otherwise outlive the failure.
        let _ = child.kill();
        let _ = child.wait();
        return Err(err);
    }

    let status = child.wait().context("failed to wait for the bulk-index plugin process")?;
    if !status.success() {
        bail!("the {} plugin's bulk index exited with {status}", manifest.language);
    }

    // This language's `language_state.bulkIndexedAt`, recorded here because
    // this is the one place that knows which language just finished;
    // `schema::record_bulk_index` is the project-wide roll-up its caller
    // writes once, after every language. The one write site of
    // `pluginFingerprint`, the digest `daemon::registry::indexer_version`
    // computes per plugin.
    store
        .step(|conn| {
            schema::record_language_bulk_indexed(
                conn,
                &manifest.language,
                Some(&plugin::fingerprint(manifest)),
            )
        })
        .with_context(|| format!("failed to record that {} was bulk-indexed", manifest.language))?;

    Ok(())
}

/// Honors [`WALK_HOLD_FILE_ENV`] and [`WALK_DELAY_ENV`]. A no-op unless one is
/// set, which is every real run.
fn hold_the_walk_open_for_tests() {
    // The file gate wins: a test that uses it wants the release to be its own
    // action, not blurred by a delay on top.
    if let Some(path) = std::env::var_os(WALK_HOLD_FILE_ENV).filter(|p| !p.is_empty()) {
        let path = std::path::PathBuf::from(path);
        eprintln!(
            "g-mesh daemon: holding the finished bulk walk until {} is removed ({WALK_HOLD_FILE_ENV})",
            path.display()
        );
        // Polled, and bounded so a test that forgets to release it fails as a
        // test timeout rather than wedging the daemon.
        let deadline = std::time::Instant::now() + std::time::Duration::from_secs(30);
        while path.exists() && std::time::Instant::now() < deadline {
            std::thread::sleep(std::time::Duration::from_millis(2));
        }
        return;
    }

    let Some(millis) = std::env::var(WALK_DELAY_ENV).ok().and_then(|v| v.trim().parse().ok()) else {
        return;
    };
    eprintln!("g-mesh daemon: holding the finished bulk walk open for {millis}ms ({WALK_DELAY_ENV})");
    std::thread::sleep(std::time::Duration::from_millis(millis));
}

/// Reads one language's whole NDJSON bulk stream to EOF into `ctx`,
/// committing it in batches, as one [`Unit::BulkWalk`]. Lets the ingestion
/// rules run without a plugin on the other end.
///
/// Adds only to `nodes`/`edges`/`skipped_lines`: `linked_imports`/
/// `linked_symbols` are [`run`]'s, computed once after every language.
pub(crate) fn ingest<R: BufRead>(reader: R, ctx: &mut WalkContext<'_>) -> Result<()> {
    let store = ctx.store;
    store.unit(Unit::BulkWalk, |store| ingest_in(reader, store, ctx))
}

fn ingest_in<R: BufRead>(reader: R, store: &mut Writer<'_>, ctx: &mut WalkContext<'_>) -> Result<()> {
    let mut batch = Diff::default();
    let mut batched = 0usize;

    for item in NdjsonReader::new(reader) {
        match item {
            Ok(BulkItem::Node(node)) => {
                let record = to_node_record(*node);
                // A `File` node is the plugin saying it parsed that file - the
                // one statement `run`'s baselines may rest on. Other kinds'
                // `filePath` need not be a file this walk read.
                if record.kind == FILE_NODE_KIND {
                    if let Some(walked) = ctx.walked_files.as_mut() {
                        walked.insert(record.file_path.clone());
                    }
                }
                batch.upsert_nodes.push(record);
                ctx.summary.nodes += 1;
            }
            Ok(BulkItem::Edge(edge)) => {
                batch.upsert_edges.push(to_edge_record(edge));
                ctx.summary.edges += 1;
            }
            Err(err) => {
                // A read failure (a broken pipe, say) repeats on every next
                // iteration, so skipping it like a malformed line would spin
                // forever: the stream is over.
                if err.downcast_ref::<std::io::Error>().is_some() {
                    return Err(err).context("failed to read the plugin's bulk-index stream");
                }
                eprintln!("g-mesh daemon: skipping malformed bulk-index line: {err:#}");
                ctx.summary.skipped_lines += 1;
                continue;
            }
        }

        if let Some(progress) = ctx.progress {
            progress.add_items_ingested(1);
        }
        batched += 1;
        if batched >= BATCH_ITEMS {
            commit(store, &mut batch, ctx)?;
            batched = 0;
        }
    }

    commit(store, &mut batch, ctx)?;
    Ok(())
}

/// Commits one batch and empties it. Embedding inference runs first, outside
/// the store; the commit and the vector store are then one step of the walk's
/// unit, so nothing inside the hold scales with inference.
fn commit(store: &mut Writer<'_>, batch: &mut Diff, ctx: &WalkContext<'_>) -> Result<()> {
    if batch.is_empty() {
        return Ok(());
    }
    let embedding = ctx.embedding;
    let computed = embedding.map(|embedding| embedding.compute(batch)).unwrap_or_default();
    store.commit_batch(batch, embedding.map(|embedding| (embedding, computed.as_slice())))?;
    *batch = Diff::default();
    Ok(())
}

/// Path whose deletion releases a batch commit holding the store open; the
/// hook fires inside [`IndexStore::commit_batch`]'s hold.
pub use crate::storage::index_store::HOLD_LOCK_FILE_ENV;

#[cfg(test)]
mod tests {
    use super::*;
    use crate::protocol::types::{
        EdgeKind, NodeKind, Position, Range, SourceTier, Visibility, WireEdge, WireNode,
    };
    use crate::storage::schema;
    use rusqlite::Connection;
    use std::io::Cursor;

    fn setup_conn() -> IndexStore {
        let conn = Connection::open_in_memory().unwrap();
        // Foreign keys on, unlike the daemon's own connection: the point of
        // several of these tests is that batching never presents SQLite with
        // an edge whose endpoints aren't in yet, and that only bites when the
        // constraint is actually enforced.
        conn.pragma_update(None, "foreign_keys", "ON").unwrap();
        schema::apply(&conn).unwrap();
        IndexStore::new(conn)
    }

    fn count(conn: &IndexStore, table: &str) -> i64 {
        conn.lock()
            .unwrap()
            .query_row(&format!("SELECT COUNT(*) FROM {table}"), [], |row| row.get(0))
            .unwrap()
    }

    fn node_line(id: &str) -> String {
        serde_json::to_string(&WireNode {
            id: id.to_string(),
            kind: NodeKind::Function,
            name: id.to_string(),
            qualified_name: id.to_string(),
            file_path: "src/a.ts".to_string(),
            range: Range { start: Position { line: 1, col: 0 }, end: Position { line: 2, col: 0 } },
            signature: None,
            visibility: Visibility::Public,
            doc_comment: None,
            language: "typescript".to_string(),
            native_kind: None,
            has_syntax_errors: false,
            declarations: None,
            container: None,
            container_parent: None,
            target: None,
        })
        .unwrap()
    }

    fn edge_line(id: &str, from: &str, to: &str) -> String {
        serde_json::to_string(&WireEdge {
            id: id.to_string(),
            from_id: from.to_string(),
            to_id: to.to_string(),
            kind: EdgeKind::Calls,
            source: SourceTier::Syntactic,
            engine: "tree-sitter".to_string(),
            resolved: false,
            to_declaration: None,
        })
        .unwrap()
    }

    fn ingest_str(stream: &str, conn: &IndexStore) -> Result<BulkIndexSummary> {
        let mut ctx = WalkContext::new(conn);
        ingest(Cursor::new(stream.as_bytes().to_vec()), &mut ctx)?;
        Ok(ctx.summary)
    }

    #[test]
    fn a_whole_stream_lands_in_sqlite() {
        let conn = setup_conn();
        let stream = format!("{}\n{}\n{}\n", node_line("n1"), node_line("n2"), edge_line("e1", "n1", "n2"));

        let summary = ingest_str(&stream, &conn).unwrap();

        assert_eq!(
            summary,
            BulkIndexSummary { nodes: 2, edges: 1, skipped_lines: 0, linked_imports: 0, linked_symbols: 0 }
        );
        assert_eq!(count(&conn, "nodes"), 2);
        assert_eq!(count(&conn, "edges"), 1);
        let name: String = conn
            .lock()
            .unwrap()
            .query_row("SELECT name FROM nodes WHERE id = 'n1'", [], |row| row.get(0))
            .unwrap();
        assert_eq!(name, "n1");
    }

    /// One unreadable line costs one symbol; refusing the whole walk over it
    /// would cost the project its entire index.
    #[test]
    fn a_malformed_line_is_counted_and_skipped_without_losing_the_rest() {
        let conn = setup_conn();
        let stream = format!("{}\nnot json at all\n{}\n", node_line("n1"), node_line("n2"));

        let summary = ingest_str(&stream, &conn).unwrap();

        assert_eq!(
            summary,
            BulkIndexSummary { nodes: 2, edges: 0, skipped_lines: 1, linked_imports: 0, linked_symbols: 0 }
        );
        assert_eq!(count(&conn, "nodes"), 2, "lines after a bad one must still be committed");
    }

    /// The batching contract: a stream longer than one batch commits every
    /// item exactly once, and a batch boundary landing between a node and an
    /// edge that points at it must not produce a dangling edge.
    #[test]
    fn a_stream_longer_than_one_batch_commits_all_of_it() {
        let conn = setup_conn();
        let total = BATCH_ITEMS + 5;
        let mut stream = String::new();
        for i in 0..total {
            stream.push_str(&node_line(&format!("n{i}")));
            stream.push('\n');
        }
        // Deliberately last, i.e. in the trailing partial batch, while its
        // endpoints were committed by the full batch before it.
        stream.push_str(&edge_line("e1", "n0", &format!("n{}", total - 1)));
        stream.push('\n');

        let summary = ingest_str(&stream, &conn).unwrap();

        assert_eq!(summary.nodes, total);
        assert_eq!(summary.edges, 1);
        assert_eq!(count(&conn, "nodes"), total as i64);
        assert_eq!(count(&conn, "edges"), 1);
    }

    #[test]
    fn an_empty_stream_is_a_no_op() {
        let conn = setup_conn();
        let summary = ingest_str("", &conn).unwrap();
        assert_eq!(summary, BulkIndexSummary::default());
        assert_eq!(count(&conn, "nodes"), 0);
    }

    /// Discovering two languages must spawn
    /// *both* plugins' one-shot `--bulk-index` mode and add their
    /// contributions together into one summary, not just walk the first (or
    /// only) one found. Each fake plugin (`daemon::test_plugin`) emits a
    /// fixed two nodes and one edge in bulk-index mode, so two languages
    /// summing to four nodes and two edges is only possible if both were
    /// actually spawned and both streams actually landed in the same index.
    #[test]
    fn discovering_two_languages_walks_both_and_sums_their_contributions() {
        let project = tempfile::tempdir().unwrap();
        let plugins = tempfile::tempdir().unwrap();
        crate::daemon::test_plugin::install(plugins.path(), "alpha", &[".alpha-src"]);
        crate::daemon::test_plugin::install(plugins.path(), "beta", &[".beta-src"]);
        let discovered = crate::daemon::manifest::discover(&[plugins.path().to_path_buf()])
            .expect("the fixture plugins must discover cleanly");
        assert_eq!(discovered.manifests.len(), 2, "both fixture languages must have been discovered");

        let conn = setup_conn();

        let summary = run(project.path(), &conn, None, &discovered).expect("the multi-language walk failed");

        assert_eq!(summary.nodes, 4, "both languages' nodes must be counted, not just one's");
        assert_eq!(summary.edges, 2, "both languages' edges must be counted, not just one's");
        assert_eq!(summary.skipped_lines, 0);
        assert_eq!(count(&conn, "nodes"), 4, "both languages' nodes must have actually been committed");
        assert_eq!(count(&conn, "edges"), 2, "both languages' edges must have actually been committed");
    }

    /// A discovery naming only one language must still walk it: the loop over
    /// `discovered.manifests` must not accidentally require more than one
    /// entry to do anything.
    #[test]
    fn a_single_discovered_language_is_walked_same_as_before_the_registry_took_over() {
        let project = tempfile::tempdir().unwrap();
        let plugins = tempfile::tempdir().unwrap();
        crate::daemon::test_plugin::install(plugins.path(), "solo", &[".solo-src"]);
        let discovered = crate::daemon::manifest::discover(&[plugins.path().to_path_buf()])
            .expect("the fixture plugin must discover cleanly");

        let conn = setup_conn();

        let summary = run(project.path(), &conn, None, &discovered).expect("the single-language walk failed");

        assert_eq!(summary.nodes, 2);
        assert_eq!(summary.edges, 1);
        assert_eq!(count(&conn, "nodes"), 2);
        assert_eq!(count(&conn, "edges"), 1);
    }

    /// An empty discovery (no plugins found at all) must not be an error -
    /// same "not fatal" convention `PluginRegistry` documents for the same
    /// case - it is simply a walk that finds nothing to spawn.
    #[test]
    fn an_empty_discovery_walks_nothing_and_is_not_an_error() {
        let project = tempfile::tempdir().unwrap();
        let conn = setup_conn();
        let discovered = DiscoveredPlugins::default();

        let summary =
            run(project.path(), &conn, None, &discovered).expect("an empty discovery must not fail");

        assert_eq!(summary, BulkIndexSummary::default());
        assert_eq!(count(&conn, "nodes"), 0);
    }
}
